docs/what-is-wrapped.md

```diff
@@ -0,0 +1,1007 @@
+trait Wrapped[T] extends Any {
+  def value: T
+  override def toString: String = value.toString
+
+  // check that this is the same Wrapper, e.g. that we are comparing `UserId` vs `UserId` and not `UserId` vs `Email`
+  def canEqual(that: Any): Boolean = this.getClass.isInstance(that)
+  // cast and compare underlying values
+  override def equals(that: Any): Boolean = canEqual(that) && this.value.equals(that.asInstanceOf[Wrapped[T]].value)
+  // hashcode takes into account wrapper class information and its content. This is done to obey equals / hashcode laws
+  override def hashCode: Int = this.getClass.hashCode + value.hashCode()
+}
+
+object Wrapped {
+  trait Companion {
+    type Type
+    type Wrapper <: Wrapped[Type]
+    
+    // we don't have to define such alias, but lest do it in order to show what this Either means
+    type ValidationResult[A] = Either[String, A]
+    // check that provided data is valid and can be used to create a wrapper
+    protected def validate(value: Type): ValidationResult[Type]
+    // unsafely creates a wrapper. Without this method we just don't know how to do it for an arbitrary type
+    protected def create(value: Type): Wrapper
+
+    // validates data and then unsafely creates wrapper
+    final def from(value: Type): ValidationResult[Wrapper] = validate(value).map(create)
+  }
+}
+
+final class UserId private (val value: String) extends AnyVal with Wrapped[String]
+object UserId extends Wrapped.Companion {
+  type Type = String
+  type Wrapper = UserId
+  
+  protected def validate(value: String): Either[String, String] = Right(value)
+  protected def create(value: String): UserId = new UserId(value)
+}
+
+final class Age private (val value: Int) extends AnyVal with Wrapped[Int]
+object Age extends Wrapped.Companion {
+  type Type = Int
+  type Wrapper = Age
+
+  protected def validate(value: Int): Either[String, Int] =
+    if (value >= 0 && value < 200) Right(value) else Left("user age should be from 0 to 200 years")
+
+  protected def create(value: Int): Age = new Age(value)
+}
+
+final class Email private (val value: String) extends AnyVal with Wrapped[String]
+object Email extends Wrapped.Companion {
+  type Type = String
+  type Wrapper = Email
+  
+  private val dummyEmailRegexp: String = "(\\w|\\d)+@(\\w|\\d)\\.(\\w|\\d)"
+  
+  protected def validate(value: String): Either[String, String] =
+    if (value.matches(dummyEmailRegexp)) Right(value) else Left(s"email must match regexp $dummyEmailRegexp")
+
+  protected def create(value: String): Email = new Email(value)
+}
+
+final case class User(
+  id: UserId,
+  age: Age,
+  email: Email
+)
+
+class EmailService {
+  def sendEmail(email: Email, content: String): Unit =
+    println(s"Sending [$content] to $email")
+}
+
+class LawsInfoService {
+  def canHaveRussianPassport(age: Age): Boolean =
+    age.value >= 14
+}
+
+class UserRepository {
+  def get(id: UserId): User = ???
+    // oops, it doesn't work anymore. We will deal with it later
```

Comment:
Don't know if that is even an issue, but that "oops, it doesn't work anymore" holds from the previous example on 735 line. Thought, that I scrolled up accidentally. Maybe something like "it still doesn't work, but wait a little and we will handle it with the most brilliant fashion" would be less confusing :)

Reply:
Agree. Changed to `still doesn't work. We will fix it soon`